test/integration/ads_integration_test.cc

```diff
@@ -0,0 +1,147 @@
+
+namespace Envoy {
+namespace {
+
+class AdsIntegrationTest : public BaseIntegrationTest,
+                           public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  AdsIntegrationTest() : BaseIntegrationTest(GetParam()) {}
+
+  void SetUp() override {
+    fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP2, version_));
+    registerPort("endpoint", fake_upstreams_.back()->localAddress()->ip()->port());
+    fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP2, version_));
+    registerPort("ads_upstream", fake_upstreams_.back()->localAddress()->ip()->port());
+    createTestServer("test/config/integration/server_ads.yaml", {"http"});
+  }
+
+  void TearDown() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void expectDiscoveryRequest(const std::string& type_url, const std::string& version) {
+    envoy::api::v2::DiscoveryRequest discovery_request;
+    ads_stream_->waitForGrpcMessage(*dispatcher_, discovery_request);
+    EXPECT_EQ(type_url, discovery_request.type_url());
+    EXPECT_TRUE(discovery_request.resource_names().empty());
+    EXPECT_EQ(version, discovery_request.version_info());
+  }
+
+  void sendDiscoveryResponse(const std::string& type_url, const Protobuf::Message& message,
+                             const std::string& version) {
+    envoy::api::v2::DiscoveryResponse discovery_response;
+    discovery_response.set_version_info(version);
+    discovery_response.set_type_url(type_url);
+    discovery_response.add_resources()->PackFrom(message);
+    ads_stream_->sendGrpcMessage(discovery_response);
+  }
+
+  envoy::api::v2::Cluster buildCluster(const std::string& name) {
+    envoy::api::v2::Cluster cluster;
+    cluster.set_name(name);
+    cluster.mutable_connect_timeout()->set_seconds(5);
+    cluster.set_type(envoy::api::v2::Cluster::EDS);
+    cluster.mutable_eds_cluster_config()->mutable_eds_config()->mutable_ads();
+    cluster.set_lb_policy(envoy::api::v2::Cluster::ROUND_ROBIN);
+    cluster.mutable_http2_protocol_options();
+    return cluster;
+  }
+
+  envoy::api::v2::ClusterLoadAssignment buildClusterLoadAssignment(const std::string& name) {
+    envoy::api::v2::ClusterLoadAssignment cluster_load_assignment;
+    cluster_load_assignment.set_cluster_name(name);
+    auto* endpoint = cluster_load_assignment.mutable_endpoints()
+                         ->Add()
+                         ->mutable_lb_endpoints()
+                         ->Add()
+                         ->mutable_endpoint();
+    auto* socket_address = endpoint->mutable_address()->mutable_socket_address();
+    socket_address->set_address(Network::Test::getLoopbackAddressString(GetParam()));
+    socket_address->set_port_value(fake_upstreams_[0]->localAddress()->ip()->port());
+    return cluster_load_assignment;
+  }
+
+  envoy::api::v2::Listener buildListener(const std::string& name, const std::string& route_config) {
```

Comment:
Related to the convo we had in meeting yesterday, and to discussions with @alyssawilk, is there any way to make this more readable? By potentially having YAML templates for listeners for most things and small helper functions that are mostly abstracted? I can look at a YAML config and immediately understand what is going on. This is pretty hard to look at...

Reply:
I don't think this function is particularly readable but I think the *use* of it is, or can be made to be.

If we 

addListener("listener1, "route1")
addRoute("route1" "cluster1" "/foo");
addCluster("cluster1", Something::Mumble::HTTP1);

Then create envoy, I think it's pretty clear we have 1 listener, which maps /foo to cluster1 which speaks HTTP
That's the sort of config I'd like to see in the long run.

Reply:
Yes I agree, that's where we need to end up, with maybe the functions themselves being based on templates that are easy to understand (or at least with good comments).

Reply:
I have made this nicer in the commit I'm working on, here's what it will look like:

```
  envoy::api::v2::Cluster buildCluster(const std::string& name) {
    auto cluster = TestUtility::parseYaml<envoy::api::v2::Cluster>(R"EOF(
      connect_timeout: 5s
      type: EDS
      eds_cluster_config: { eds_config: { ads: {} } }
      lb_policy: ROUND_ROBIN
      http2_protocol_options: {}
    )EOF");
    cluster.set_name(name);
    return cluster;
  }
```